### DISCUSSION TITLE
Nuxt3 SSR hydration changes URL during hydration behind a cloudfront cache

### DISCUSSION THREAD

**Discussion by ricketyboo** [nuxt/nuxt#23511] [Questions] [answered]
We just did a migration from Nuxt2 to Nuxt3.
Site has been running in AWS behind a cloudfront layer with cache keys based on a couple of query parameters that actually drive functional change. 

As an example `/page`, `/page?category=foo` and `/page?category=bar` will return very different content with an expensive database call for each. We have `category` configured as a cache key for cloudfront so if it is set we get cached results for the `foo` variant and the `bar` variant, and if it's not set we get the default variant.

The url quite often has other, unimportant to the rendering of the params such as `utm_campaign` or `gclid` and `fbclid` for ad tracking. These are only used by the client side through tag manager scripts and are irrelevant to the nuxt ssr side.
Eg `/page?category=foo&utm_campaign=marketing2023&gclid=123456789`

In Nuxt v2 this worked fine. SSR renders, cloudfront caches, and everything is great.

In Nuxt v3 however we have discovered that the hydration step of the SSR render is storing the path + query params in the __NUXT_DATA__ payload, and on hydration alters the client URL to match whatever the SSR thought the URL was at render time. This is now causing us problems in that if someone comes in on some url with query parameters, everyone who comes in after them has the URL changed to whatever the initial generation was triggered with

eg 
cache key for query param category
user 1 lands on `/page?utm_campaign=20dollarsoff` which gets cached as default `/page` response
user 2-10 navigates to `/page` and the url is changed during hydration to `/page?utm_campaign=20dollarsoff` which means client side scripts now see this param and breaks attribution
user 11 visits `/page?category=foo&utm_campaign=100dollarsoff` which is cached as a variant for page + category
user 12-20 visits `/page?category=foo` and url changes during hydration to `/page?category=foo&utm_campaign=100dollarsoff` and now client side scripts see this param and breaks attribution

We can sort of hide this issue by enabling all query params as cache key indexes in cloudfront, but now almost every url is a unique cache index due to thinks like google and facebook click ids, and our cache is basically useless with 70% of requests being cache misses.

Is there anyway to make nuxt3 work as nuxt2 did and not alter the url during hydration?

**Comment by ricketyboo** [comment]
After much tinkering with renderer.js, payload.js, revive-payload.js, and router.js I'm arriving at the belief that the full URL is now just inextricably linked to the SSR state. If the url in the payload ever differs from the client url at all, hydration panics and starts refiring all the BE requests in the client side which is a not desirable.

I'd love to be advised that I've missed something obvious, I don't understand how no-one else seems to have faced this issue. Do people just not bother with cache and let everything hit the server for every request? Unfortunately that's not an option here as the server behind our nuxt instance can't handle load, and content doesn't change enough to warrant requesting data again for every request.

Next option for anyone else who might stumble on this in the future is to try putting a caching layer between Nuxt and our backend service that it talks to, and remove the cache between users and vue/nuxt. This is less ideal as we're now going to be upping the amount of time and $ spend on a live server. Also suspect there's going to be increased TTFB and load times as we will have to wait for Nuxt to open connection, get a (cached) response, close connection, process, return content, where previously this only happened when cache expiry was triggered (rarely).

I still hope someone in the project might see this and have an easy idea to support this idea of deployment behind a content cache, as I imagine other people would be using this architecture as well.

**Comment by dreitzner** [comment]
We are currently investigating the same issue.

We use cloudflare workers as a caching layer. It might be possible to mangle the response, but I think that would be suboptimal.

**Comment by igor322** [comment, marked as answer]
We had the same problem using cloudfront cache and the query params were being changed on client side. Your comment helped us understand what was happening, so thank you. We found a workaround for it for the moment, taken from the issue #23153. We added in our layout:
`await useAsyncData('fixQueryParams', 
	async (nuxtApp) => {
		delete nuxtApp?.payload?.path;
	}
);
`

**Reply by ricketyboo** [reply]
Oh wow, thanks so much for that. I don't know how I missed that original thread, I tried so many keywords in search and just missed it completely.

Appreciate the answer and the link!
